sqlglot/dialects/dialect.py

```diff
@@ -223,26 +253,73 @@ class Dialect(metaclass=_Dialect):
 
     INVERSE_ESCAPE_SEQUENCES: t.Dict[str, str] = {}
 
-    def __eq__(self, other: t.Any) -> bool:
-        return type(self) == other
+    @t.overload
+    @classmethod
+    def get_or_raise(cls, dialect: DialectType) -> Dialect:
+        ...
 
-    def __hash__(self) -> int:
-        return hash(type(self))
+    @t.overload
+    @classmethod
+    def get_or_raise(cls, dialect: DialectType, instance: Literal[True]) -> Dialect:
+        ...
 
+    @t.overload
     @classmethod
-    def get_or_raise(cls, dialect: DialectType) -> t.Type[Dialect]:
+    def get_or_raise(cls, dialect: DialectType, instance: Literal[False]) -> t.Type[Dialect]:
```

Comment:
why would ever want a non instance? let's just make it simpler and not have instance flag

Reply:
I don't have a strong opinion on this, I just started off with the `instance` flag to make sure we've discussed about the performance tradeoffs of always instantiating vs returning a type (previous behavior)

Reply:
yea but in most places we instantiate already